source/validate_id.cpp

```diff
@@ -2495,6 +2518,75 @@ bool idUsage::isValid(const spv_instruction_t* inst) {
 #undef TODO
 #undef CASE
 }
+bool idUsage::AreInterchangeableSructs(const libspirv::Instruction* type1,
+                                       const libspirv::Instruction* type2) {
+  if (type1->opcode() != SpvOpTypeStruct) {
+    return false;
+  }
+  if (type2->opcode() != SpvOpTypeStruct) {
+    return false;
+  }
+
+  if (!HaveSameOpTypeStruct(type1, type2)) return false;
+
+  return HaveSameLayoutDecorations(type1, type2);
+}
+bool idUsage::HaveSameOpTypeStruct(const libspirv::Instruction* type1,
+                                   const libspirv::Instruction* type2) {
+  assert(type1->opcode() == SpvOpTypeStruct &&
+      "type1 must be and OpTypeStruct instruction.");
+  assert(type2->opcode() == SpvOpTypeStruct &&
+      "type2 must be and OpTypeStruct instruction.");
+  const auto& type1_operands = type1->operands();
+  const auto& type2_operands = type2->operands();
+  if (type1_operands.size() != type2_operands.size()) {
+    return false;
+  }
+
+  for (size_t operand = 2; operand < type1_operands.size(); ++operand) {
+    if (type1->word(operand) != type2->word(operand)) {
+      auto def1 = module_.FindDef(type1->word(operand));
+      auto def2 = module_.FindDef(type2->word(operand));
+      if (!AreInterchangeableSructs(def1, def2)) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+bool idUsage::HaveSameLayoutDecorations(const libspirv::Instruction* type1,
+                                        const libspirv::Instruction* type2) {
+  assert(type1->opcode() == SpvOpTypeStruct &&
+      "type1 must be and OpTypeStruct instruction.");
+  assert(type2->opcode() == SpvOpTypeStruct &&
+      "type2 must be and OpTypeStruct instruction.");
+  const std::vector<Decoration>& type1_decorations =
+      module_.id_decorations(type1->id());
+  const std::vector<Decoration>& type2_decorations =
+      module_.id_decorations(type2->id());
+
+  for (const Decoration& decoration : type1_decorations) {
+    switch (decoration.dec_type()) {
+      case SpvDecorationArrayStride:
+      case SpvDecorationOffset:
+      case SpvDecorationMatrixStride:
+      case SpvDecorationRowMajor:
+      case SpvDecorationColMajor:
+        // Since these effect the layout of the struct, they must be present in
+        // both structs.
+        if (std::find(type2_decorations.begin(), type2_decorations.end(),
+                      decoration) == type2_decorations.end()) {
+          return false;
```

Comment:
Can type2 have decorations which type1 doesn't have? (question out if ignorance)

Reply:
I've added a comment explaining this.  I don't think they can in a legal program.  The layout has to be completely determined.  For example, every member in the struct has to have an offset.  If the offset for the first member in type1 matches the offset for the first member in type 2, then type 2 cannot have a different giving the first member a different offset.  This same reasoning applies to all of the possible decorations.

Reply:
The Vulkan environment spec requires that layout decorations be complete.  (e.g. all structs have offsets for all members, all arrays have strides, etc.)  So if the member type is the of the same optype* then they'll also have layout decorations.
I expect OpenGL does the same.

Reply:
I apologize, I answered incorrectly because I forgot a key qualification.  @atgoo is right to question this.

_When objects are in UniformConstant, Uniform, PushConstant, and StorageBuffer storage classes_, then their layout must be explicit.    But I think this code could come across cases where type1 is a struct value that was never associated with an object on of those storage classes.  And so you can get the case where type2 has layout constraints but type1 does not.